File: packages/expo-gateflow/ios/GateFlowExpoModule.swift
import ExpoModulesCore

/**
 * Native module for GateFlow A/B experimentation.
 * Bridges the GateFlow SDK to React Native via Expo Modules Core.
 */
public class GateFlowExpoModule: Module {
  private var bridgeAdapter: GateFlowBridgeAdapter?

  public func definition() -> ModuleDefinition {
    Name("GateFlowExpo")

    Events(
      "onConfigSuccess",
      "onConfigFail",
      "onConfigRefresh",
      "onUserIdentified",
      "onUserReset",
      "onExperimentEvaluated",
      "onEventTracked",
      "onUserAttributesUpdated"
    )

    Function("configure") { (apiUrl: String, apiKey: String, options: [String: Any]?) async throws in
      let adapter = try getOrCreateAdapter()
      try await adapter.configure(apiUrl: apiUrl, apiKey: apiKey, options: options)
    }

    Function("identify") { (userId: String, userAttrs: [String: Any]?) async throws in
      let adapter = try getOrCreateAdapter()
      try await adapter.identify(userId: userId, userAttrs: userAttrs)
    }

    Function("reset") { () async throws in
      let adapter = try getOrCreateAdapter()
      try await adapter.reset()
    }

    Function("evaluateExperiment") { (placementKey: String, params: [String: Any]?, handlerId: String?) async -> [String: Any] in
      let adapter = try getOrCreateAdapter()
      return try await adapter.evaluateExperiment(
        placementKey: placementKey,
        params: params,
        handlerId: handlerId
      )
    }

    Function("getExperiment") { (key: String) async -> [String: Any]? in
      let adapter = try getOrCreateAdapter()
      return try await adapter.getExperiment(key: key)
    }

    Function("getAllExperiments") { () async -> [[String: Any]] in
      let adapter = try getOrCreateAdapter()
      return try await adapter.getAllExperiments()
    }

    Function("trackEvent") { (event: [String: Any]) async throws in
      let adapter = try getOrCreateAdapter()
      try await adapter.trackEvent(event: event)
    }

    Function("setUserAttributes") { (attrs: [String: Any]) async throws in
      let adapter = try getOrCreateAdapter()
      try await adapter.setUserAttributes(attrs: attrs)
    }

    Function("getUserAttributes") { () async -> [String: Any] in
      let adapter = try getOrCreateAdapter()
      return try await adapter.getUserAttributes()
    }

    Function("preloadExperiments") { (placementKeys: [String]) async throws in
      let adapter = try getOrCreateAdapter()
      try await adapter.preloadExperiments(placementKeys: placementKeys)
    }
  }

  private func getOrCreateAdapter() throws -> GateFlowBridgeAdapter {
    if let adapter = bridgeAdapter {
      return adapter
    }
    let adapter = GateFlowBridgeAdapter(sendEvent: { [weak self] eventName, payload in
      self?.sendEvent(eventName, payload)
    })
    bridgeAdapter = adapter
    return adapter
  }
}

File: packages/expo-gateflow/android/src/main/java/com/gateflow/expo/GateFlowExpoModule.kt
package com.gateflow.expo

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.types.Any

/**
 * Native module for GateFlow A/B experimentation on Android.
 * Bridges the GateFlow SDK to React Native via Expo Modules Core.
 */
class GateFlowExpoModule : Module() {
  private var bridgeAdapter: GateFlowBridgeAdapter? = null

  override fun definition() = ModuleDefinition {
    Name("GateFlowExpo")

    Events(
      "onConfigSuccess",
      "onConfigFail",
      "onConfigRefresh",
      "onUserIdentified",
      "onUserReset",
      "onExperimentEvaluated",
      "onEventTracked",
      "onUserAttributesUpdated"
    )

    AsyncFunction("configure") Coroutine { apiUrl: String, apiKey: String, options: Map<String, Any>? ->
      getOrCreateAdapter().configure(apiUrl, apiKey, options)
    }

    AsyncFunction("identify") Coroutine { userId: String, userAttrs: Map<String, Any>? ->
      getOrCreateAdapter().identify(userId, userAttrs)
    }

    AsyncFunction("reset") Coroutine {
      getOrCreateAdapter().reset()
    }

    AsyncFunction("evaluateExperiment") Coroutine { placementKey: String, params: Map<String, Any>?, handlerId: String? ->
      getOrCreateAdapter().evaluateExperiment(placementKey, params, handlerId)
    }

    AsyncFunction("getExperiment") Coroutine { key: String ->
      getOrCreateAdapter().getExperiment(key)
    }

    AsyncFunction("getAllExperiments") Coroutine {
      getOrCreateAdapter().getAllExperiments()
    }

    AsyncFunction("trackEvent") Coroutine { event: Map<String, Any> ->
      getOrCreateAdapter().trackEvent(event)
    }

    AsyncFunction("setUserAttributes") Coroutine { attrs: Map<String, Any> ->
      getOrCreateAdapter().setUserAttributes(attrs)
    }

    AsyncFunction("getUserAttributes") Coroutine {
      getOrCreateAdapter().getUserAttributes()
    }

    AsyncFunction("preloadExperiments") Coroutine { placementKeys: List<String> ->
      getOrCreateAdapter().preloadExperiments(placementKeys)
    }
  }

  private fun getOrCreateAdapter(): GateFlowBridgeAdapter {
    if (bridgeAdapter != null) return bridgeAdapter!!
    val adapter = GateFlowBridgeAdapter(
      context = appContext.reactContext ?: throw IllegalStateException("ReactContext not available"),
      sendEvent = { eventName, payload -> sendEvent(eventName, payload) }
    )
    bridgeAdapter = adapter
    return adapter
  }
}
